File: packages/tools/src/tools/edit.ts
import { readFile, writeFile } from "node:fs/promises";
import type { ToolDefinition, ToolResult } from "@forge/types";

export const editTool: ToolDefinition = {
  name: "Edit",
  description:
    "Performs exact string replacements in a file. The old_string must be unique " +
    "in the file (unless replace_all is true). Returns an error if old_string is " +
    "not found or appears multiple times without replace_all enabled.",
  inputSchema: {
    type: "object",
    properties: {
      file_path: {
        type: "string",
        description: "The absolute path to the file to edit",
      },
      old_string: {
        type: "string",
        description: "The text to replace (must be exact match)",
      },
      new_string: {
        type: "string",
        description: "The text to replace it with",
      },
      replace_all: {
        type: "boolean",
        description:
          "Replace all occurrences instead of requiring uniqueness (default: false)",
        default: false,
      },
    },
    required: ["file_path", "old_string", "new_string"],
  },
  annotations: { readOnly: false },
  handler: async (input): Promise<ToolResult> => {
    const filePath = input.file_path as string;
    const oldString = input.old_string as string;
    const newString = input.new_string as string;
    const replaceAll = (input.replace_all as boolean | undefined) ?? false;

    try {
      // Read file
      const content = await readFile(filePath, "utf-8");

      // Count occurrences
      const occurrences = countOccurrences(content, oldString);

      if (occurrences === 0) {
        return {
          content: [
            {
              type: "text",
              text: `String not found in file: "${oldString}"`,
            },
          ],
          isError: true,
        };
      }

      if (occurrences > 1 && !replaceAll) {
        return {
          content: [
            {
              type: "text",
              text:
                `String appears ${occurrences} times in file and is not unique. ` +
                `Use replace_all: true to replace all occurrences.`,
            },
          ],
          isError: true,
        };
      }

      // Perform replacement
      const newContent = replaceAll
        ? content.split(oldString).join(newString)
        : content.replace(oldString, newString);

      // Write back
      await writeFile(filePath, newContent, "utf-8");

      const message =
        occurrences === 1
          ? `Successfully replaced 1 occurrence in ${filePath}`
          : `Successfully replaced ${occurrences} occurrences in ${filePath}`;

      return {
        content: [{ type: "text", text: message }],
      };
    } catch (err) {
      const error = err as NodeJS.ErrnoException;
      if (error.code === "ENOENT") {
        return {
          content: [
            {
              type: "text",
              text: `File does not exist: ${filePath}`,
            },
          ],
          isError: true,
        };
      }
      return {
        content: [
          {
            type: "text",
            text: `Failed to edit file: ${error.message}`,
          },
        ],
        isError: true,
      };
    }
  },
};

function countOccurrences(content: string, search: string): number {
  if (search.length === 0) return 0;
  let count = 0;
  let pos = 0;
  while ((pos = content.indexOf(search, pos)) !== -1) {
    count++;
    pos += search.length;
  }
  return count;
}

File: internal/tools/edit.go
package tools

import (
	"fmt"
	"os"
	"strings"

	"github.com/jelmersnoeck/forge/internal/types"
)

// EditTool returns the Edit tool definition.
func EditTool() types.ToolDefinition {
	return types.ToolDefinition{
		Name:        "Edit",
		Description: "Performs exact string replacement in a file. old_string must be unique unless replace_all is true.",
		InputSchema: map[string]any{
			"type": "object",
			"properties": map[string]any{
				"file_path": map[string]any{
					"type":        "string",
					"description": "Absolute path to the file",
				},
				"old_string": map[string]any{
					"type":        "string",
					"description": "Text to replace",
				},
				"new_string": map[string]any{
					"type":        "string",
					"description": "Replacement text",
				},
				"replace_all": map[string]any{
					"type":        "boolean",
					"description": "Replace all occurrences (default false)",
				},
				"disable_fuzzy": map[string]any{
					"type":        "boolean",
					"description": "Force exact matching only (default false)",
				},
			},
			"required": []string{"file_path", "old_string", "new_string"},
		},
		Handler:     editHandler,
		ReadOnly:    false,
		Destructive: false,
	}
}

func editHandler(input map[string]any, ctx types.ToolContext) (types.ToolResult, error) {
	filePath, err := requireString(input, "file_path")
	if err != nil {
		return types.ToolResult{IsError: true}, err
	}
	oldString, err := requireString(input, "old_string")
	if err != nil {
		return types.ToolResult{IsError: true}, err
	}
	newString, ok := input["new_string"].(string)
	if !ok {
		return types.ToolResult{IsError: true}, fmt.Errorf("new_string is required")
	}

	if isEnvFile(filePath) {
		return envFileError(filePath), nil
	}

	replaceAll := optionalBool(input, "replace_all", false)
	disableFuzzy := optionalBool(input, "disable_fuzzy", false)

	data, err := os.ReadFile(filePath)
	if err != nil {
		if os.IsNotExist(err) {
			return errResultf("file not found: %s", filePath)
		}
		return types.ToolResult{IsError: true}, err
	}

	content := string(data)
	ranges, strat, ok := findMatches(content, oldString, disableFuzzy)

	switch {
	case !ok:
		diag := nearMissDiagnostic(content, oldString)
		if diag != "" {
			return errResultf("%s\n\nold_string not found in %s", diag, filePath)
		}
		return errResultf("old_string not found in %s", filePath)
	case len(ranges) > 1 && !replaceAll:
		return errResultf("old_string appears %d times in %s; use replace_all: true to replace all occurrences", len(ranges), filePath)
	}

	if !replaceAll {
		ranges = ranges[:1]
	}

	newContent := spliceRanges(content, ranges, newString)

	if err := atomicWrite(filePath, []byte(newContent), targetPerm(filePath, 0644)); err != nil {
		return errResultf("failed to write file: %v", err)
	}

	ctx.ReadState.Delete(filePath)

	suffix := ""
	if strat == matchWhitespace {
		suffix = " (whitespace-normalized match)"
	}

	return textResult(fmt.Sprintf("replaced %d occurrence(s) in %s%s", len(ranges), filePath, suffix)), nil
}

// spliceRanges replaces each byte range in content (ascending, non-overlapping)
// with replacement and returns the result.
func spliceRanges(content string, ranges [][2]int, replacement string) string {
	var b strings.Builder
	prev := 0
	for _, r := range ranges {
		b.WriteString(content[prev:r[0]])
		b.WriteString(replacement)
		prev = r[1]
	}
	b.WriteString(content[prev:])
	return b.String()
}
